### ISSUE TITLE
Metaschema Compose Pipeline's Prune Unused Definition Step Not Properly Pruning Unused Definitions

### ISSUE THREAD

**Issue by aj-stein-nist** [usnistgov/metaschema#198]
# Describe the bug

As a developer, in order to have the finalized artifacts (JSON schemas, XML schemas, documentation) from the Metaschema composition pipeline process have only used definitions for fields, flags, and assemblies, I want the pipeline to properly report when it detects any unused definition and not include those in resulting schemas.

# Who is the bug affecting?

NIST OSCAL developers and other OSCAL developers defining custom models in Metaschema (outside those provided by the NIST OSCAL Team).

# What is affected by this bug?

Detection and reporting of unused Metaschema definitions that need to be pruned.

# When does this occur?

Consistently when unused definitions are present. Per https://github.com/usnistgov/metaschema/issues/180#issuecomment-1078001229, we had not seen this in current Metaschema defs of OSCAL (at least around the commits of https://github.com/usnistgov/OSCAL/commit/bfcd0082bd4c08e16c7d5f5515adc5c858e8308e) because there are no unused definitions to prune.

# How do we replicate the issue?

1. Take the sample file.
2. Apply the XSLT transform `metaschema-prune-unused-definitions4.xsl` in https://github.com/usnistgov/metaschema/issues/180#issuecomment-1078001229 or the current one as with the PR branch in usnistgov/metaschema#196 or prior to the workaround.
3. Observe unused definitions are not properly pruned.

<details>
<summary>Example Edge Case Document Instance</summary>

```xml<?xml version="1.0" encoding="UTF-8"?>
<?xml-stylesheet type="text/css" href="../../../toolchains/xslt-M4/lib/metaschema-author.css"?>
<?xml-model href="../validate/metaschema-composition-check.sch" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<METASCHEMA xmlns="http://csrc.nist.gov/ns/oscal/metaschema/1.0">
    <schema-name>Nested Reference Test</schema-name>
    <schema-version>1.0</schema-version>
    <short-name>everything</short-name>
    <namespace>http://csrc.nist.gov/metaschema/ns/nested</namespace>
    <json-base-uri>http://csrc.nist.gov/metaschema/ns/nested</json-base-uri>
    
    <remarks>
        <p>If either 'unused' definition appears in a schema, we have a bug.</p>
    </remarks>

    <define-assembly name="box">
        <formal-name>BOX</formal-name>
        <description>BOX</description>
        <root-name>BOX</root-name>
        <model/>
    </define-assembly>
    
    <define-assembly name="unused-assembly">
        <formal-name>Unused</formal-name>
        <description>BOX</description>
        <model>
            <field ref="unused-field"/>
        </model>
    </define-assembly>
    
    <define-field name="unused-field">
        <formal-name>Unused field</formal-name>
        <description>Unused field</description>
    </define-field>
    
</METASCHEMA>
```
</details>

<details>
<summary>Transform</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
    xmlns:xs="http://www.w3.org/2001/XMLSchema"
    xmlns:math="http://www.w3.org/2005/xpath-functions/math"
    xmlns:map="http://www.w3.org/2005/xpath-functions/map"
    xmlns:m="http://csrc.nist.gov/ns/oscal/metaschema/1.0"
    xmlns="http://csrc.nist.gov/ns/oscal/metaschema/1.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xpath-default-namespace="http://csrc.nist.gov/ns/oscal/metaschema/1.0"
    exclude-result-prefixes="xs math m xsi" version="3.0"
    xmlns:fn="http://www.example.com/fn">

    <xsl:output indent="yes"/>

    <xsl:strip-space
        elements="METASCHEMA define-flag define-field define-assembly remarks model choice"/>

    <xsl:variable name="show-warnings" as="xs:string">yes</xsl:variable>
    <xsl:variable name="verbose" select="lower-case($show-warnings) = ('yes', 'y', '1', 'true')"/>

    <!--<xsl:key name="global-assembly-definition" match="METASCHEMA/define-assembly" use="@_key-name"/>
    <xsl:key name="global-field-definition"    match="METASCHEMA/define-field"    use="@_key-name"/>-->
    
    
    <!-- ====== ====== ====== ====== ====== ====== ====== ====== ====== ====== ====== ====== -->
    <!-- Pass Three: filter definitions (2) - keep only top-level definitions that are actually
         called by references in the models descending from assemblies with root-name.
         
         Note that we ignore flag definitions, which can stay,
         since all flag definitions are rewritten into local declarations,
         so all top-level flag definitions will be dropped in any case.
    -->
    
<!-- This doesn't actually filter all unused definitions, because it doesn't actually traverse the tree of used definitions - if a definition is referenced from an unused definition, it is considered referenced  -->
    
    <xsl:mode on-no-match="shallow-copy"/>

    <!--<xsl:variable name="root-assembly-definitions" select="//METASCHEMA/define-assembly[exists(root-name)]"/>-->

   <xsl:template match="/METASCHEMA">
        <xsl:copy>
            <xsl:copy-of select="@*"/>
            <!--<xsl:if test="not(@abstract='yes') and empty($root-assembly-definitions)">
                <EXCEPTION problem-type="missing-root">No root found in this metaschema composition.</EXCEPTION>
            </xsl:if>-->
            
            <xsl:apply-templates/>
            
        </xsl:copy>
    </xsl:template>
    
    <!-- 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 -->    

    <xsl:key name="refs-for-definition-key" match="m:assembly | m:field | m:flag"
        use="(name(.), @_key-ref) => string-join('#')"/>
    
    
    <xsl:function name="m:key-for-definition" as="xs:string">
        <xsl:param name="def" as="element()"/>
        <xsl:value-of select="$def/string((substring-after(name(.),'define-'), @_key-name) => string-join('#'))"/>
    </xsl:function>
    
    <xsl:template match="METASCHEMA[not(@abstract='yes')]/define-assembly[empty( m:root-name | key('refs-for-definition-key',m:key-for-definition(.)) )]">
        <xsl:call-template name="warning">
            <!-- since we can detect unused definitions a better way, this can be
                 silenced and/or rendered as a comment -->
            <xsl:with-param name="type">unused-definition</xsl:with-param>
            <xsl:with-param name="msg" expand-text="true">REMOVING unused assembly definition for '{ @name }' from { ancestor::METASCHEMA[1]/@module
                }</xsl:with-param>
        </xsl:call-template>
    </xsl:template>
    
    <xsl:template match="METASCHEMA[not(@abstract='yes')]/define-field[empty(key('refs-for-definition-key',m:key-for-definition(.)) )]">
        <xsl:call-template name="warning">
            <!-- since we can detect unused definitions a better way, this can be
                 silenced and/or rendered as a comment -->
            <xsl:with-param name="type">unused-definition</xsl:with-param>
            <xsl:with-param name="msg" expand-text="true">REMOVING unused field definition for '{ @name }' from { ancestor::METASCHEMA[1]/@module
                }</xsl:with-param>
        </xsl:call-template>
    </xsl:template>
    
    <xsl:template match="METASCHEMA[not(@abstract='yes')]/define-flag[empty(key('refs-for-definition-key',m:key-for-definition(.)) )]">
        <xsl:call-template name="warning">
            <!-- since we can detect unused definitions a better way, this can be
                 silenced and/or rendered as a comment -->
            <xsl:with-param name="type">unused-definition</xsl:with-param>
            <xsl:with-param name="msg" expand-text="true">REMOVING unused flag definition for '{ @name }' from { ancestor::METASCHEMA[1]/@module
                }</xsl:with-param>
        </xsl:call-template>
    </xsl:template>
    
    <xsl:template name="warning">
        <xsl:param name="msg"/>
        <xsl:param name="type">warning</xsl:param>
        <xsl:if test="$verbose">
            <EXCEPTION problem-type="{ $type }">
                <xsl:copy-of select="$msg"/>
            </EXCEPTION>
        </xsl:if>
    </xsl:template>
    
</xsl:stylesheet>
```
</details>

<details>
<summary>Resulting "Pruned" Document Instance after Transform, with Unused Definitions Present</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-stylesheet type="text/css" href="../../../toolchains/xslt-M4/lib/metaschema-author.css"?>
<?xml-model href="../validate/metaschema-composition-check.sch" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<METASCHEMA xmlns="http://csrc.nist.gov/ns/oscal/metaschema/1.0">
   <schema-name>Nested Reference Test</schema-name>
   <schema-version>1.0</schema-version>
   <short-name>everything</short-name>
   <namespace>http://csrc.nist.gov/metaschema/ns/nested</namespace>
   <json-base-uri>http://csrc.nist.gov/metaschema/ns/nested</json-base-uri>
   <remarks>
      <p>If either 'unused' definition appears in a schema, we have a bug.</p>
   </remarks>
   <define-assembly name="box">
      <formal-name>BOX</formal-name>
      <description>BOX</description>
      <root-name>BOX</root-name>
      <model/>
   </define-assembly>
   <EXCEPTION xmlns:fn="http://www.example.com/fn"
              xmlns:map="http://www.w3.org/2005/xpath-functions/map"
              problem-type="unused-definition">REMOVING unused assembly definition for 'unused-assembly' from </EXCEPTION>
   <define-field name="unused-field">
      <formal-name>Unused field</formal-name>
      <description>Unused field</description>
   </define-field>
</METASCHEMA>
```
</details>

If applicable, add screenshots to help explain your problem.}

# Expected behavior (i.e. solution)

<details>
<summary>Resulting "Pruned" Document Instance after Transform, Happy Path with Unused Defs Pruned</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-stylesheet type="text/css" href="../../../toolchains/xslt-M4/lib/metaschema-author.css"?>
<?xml-model href="../validate/metaschema-composition-check.sch" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<METASCHEMA xmlns="http://csrc.nist.gov/ns/oscal/metaschema/1.0">
   <schema-name>Nested Reference Test</schema-name>
   <schema-version>1.0</schema-version>
   <short-name>everything</short-name>
   <namespace>http://csrc.nist.gov/metaschema/ns/nested</namespace>
   <json-base-uri>http://csrc.nist.gov/metaschema/ns/nested</json-base-uri>
   <remarks>
      <p>If either 'unused' definition appears in a schema, we have a bug.</p>
   </remarks>
   <define-assembly name="box">
      <formal-name>BOX</formal-name>
      <description>BOX</description>
      <root-name>BOX</root-name>
      <model/>
   </define-assembly>
</METASCHEMA>
```
</details>

# Other Comments

{Add any other context about the problem here.}

**Comment by aj-stein-nist** [comment]
@wendellpiez, sorry for the delay. I talked with @david-waltermire-nist and he would like the XSpec tests built into the CI/CD pipeline like [we do with the consuming OSCAL repo and its test](https://github.com/usnistgov/OSCAL/blob/bfcd0082bd4c08e16c7d5f5515adc5c858e8308e/.github/workflows/workflow-generate-metaschema-resources.yml#L114-L118). We don't have CI/CD in GitHub Actions, but CircleCI, the same approach should work and I can definitely help on that angle.

Also we don't have acceptance criteria in bug reports, but I am sure we get around to it in the PR.

**Comment by aj-stein-nist** [comment]
@wendellpiez is close to completing #199 to stabilize tests relating to the code fixes that will close #196 and #198. After talking with @david-waltermire-nist and Wendell today, we are going to focus on just resolving the immediate bug, and made us realize two related but separate streams of work.

- Fix existing integration-level (end-to-end, e2e) tests that test all Metaschema composition phases, establish baseline for MVP unit tests
- Update CI/CD for the Metaschema repository to automate the above tests, indicating on pull requests they are not recommended for approval/merge
